flink-ml-lib/src/main/java/org/apache/flink/ml/common/allreduce/AllReduceUtils.java

```diff
@@ -0,0 +1,298 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.common.allreduce;
+
+import org.apache.flink.api.common.typeinfo.BasicTypeInfo;
+import org.apache.flink.api.common.typeinfo.PrimitiveArrayTypeInfo;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.tuple.Tuple3;
+import org.apache.flink.api.java.tuple.Tuple4;
+import org.apache.flink.api.java.typeutils.TupleTypeInfo;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.BoundedOneInput;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.util.Preconditions;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Applies all-reduce on a DataStream where each partition contains only one double array.
+ *
+ * <p>AllReduce is a communication primitive widely used in MPI. In this implementation, all workers
+ * do reduce on a partition of the whole data and they all get the final reduce result. In detail,
+ * we split each double array into pieces of fixed size buffer (4KB by default) and let each subtask
+ * handle several pieces.
+ *
+ * <p>There're mainly three stages:
+ * <li>1. All workers send their partial data to other workers for reduce.
+ * <li>2. All workers do reduce on all data it received and then send partial results to others.
+ * <li>3. All workers merge partial results into final result.
+ */
+public class AllReduceUtils {
+
+    private static final int TRANSFER_BUFFER_SIZE = 1024 * 4;
+
+    public static DataStream<double[]> allReduce(DataStream<double[]> input) {
```

Comment:
Could you add Java doc here explaining the relationship between the inputs and outputs of this method?

Reply:
Hi Dong, Thanks for the reivew.

The Java doc is updated.